gem/lib/textmate_grammar/grammar.rb

```diff
@@ -0,0 +1,560 @@
+    #   the current git commit as the version
+    # @option keys [Array] :patterns ([]) ignored, will be replaced with the initial context
+    # @option keys [Hash] :repository ({}) ignored, will be replaced by the generated rules
+    # @option keys all remaining options will be copied to the grammar without change
+    #
+    def initialize(keys)
+        required_keys = [:name, :scope_name]
+        unless required_keys & keys.keys == required_keys
+            puts "Missing one or more of the required grammar keys"
+            puts "Missing: #{required_keys - (required_keys & keys.keys)}"
+            puts "The required grammar keys are: #{required_keys}"
+            raise "See above error"
+        end
+
+        @name = keys[:name]
+        @scope_name = keys[:scope_name]
+        @repository = {}
+
+        keys.delete :name
+        keys.delete :scope_name
+
+        # auto versioning, when save_to is called grab the latest git commit or "" if not
+        # a git repo
+        keys[:version] ||= :auto
+        @keys = keys
+        return if @scope_name == "export" || @scope_name.start_with?("source.", "text.")
+
+        puts "Warning: grammar scope name should start with `source.' or `text.'"
+        puts "Examples: source.cpp text.html text.html.markdown source.js.regexp"
+    end
+
+    #
+    # Access a pattern in the grammar
+    #
+    # @param [Symbol] key The key the pattern is stored in
+    #
+    # @return [PatternBase, Symbol, Array<PatternBase, Symbol>] The stored pattern
+    #
+    def [](key)
+        @repository.fetch(key, PlaceholderPattern.new(key))
+    end
+
+    #
+    # Store a pattern
+    #
+    # A pattern must be stored in the grammar for it to appear in the final grammar
+    #
+    # The special key :$initial_context is the pattern that will be matched at the
+    # beginning of the document or whenever the root of the grammar is to be matched
+    #
+    # @param [Symbol] key The key to store the pattern in
+    # @param [PatternBase, Symbol, Array<PatternBase, Symbol>] value the pattern to store
+    #
+    # @return [PatternBase, Symbol, Array<PatternBase, Symbol>] the stored pattern
+    #
+    def []=(key, value)
+        unless key.is_a? Symbol
+            raise "Use symbols not strings" unless key.is_a? Symbol
+        end
+
+        if key.to_s.start_with?("$") && !([:$initial_context, :$base, :$self].include? key)
+            puts "#{key} is not a valid repository name"
+            puts "repository names starting with $ are reserved"
+            raise "See above error"
+        end
+
+        if key.to_s == "repository"
+            puts "#{key} is not a valid repository name"
+            puts "the name 'repository' is a reserved name"
+            raise "See above error"
+        end
+
+        # ensure array is flat and only contains patterns
+        if value.is_a? Array
+            value = value.flatten.map do |item|
+                next item if item.is_a? Symbol
```

Comment:
Are all Symbols skipped here?

Reply:
Symbols and patterns are kept as is, anything else is coaxed into a pattern.